File: csharp/examples/ProducerFifoMessageExample.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using NLog;
using Org.Apache.Rocketmq;

namespace examples
{
    static class ProducerFifoMessageExample
    {
        private static readonly Logger Logger = MqLogManager.Instance.GetCurrentClassLogger();

        internal static async Task QuickStart()
        {
            string accessKey = "yourAccessKey";
            string secretKey = "yourSecretKey";
            // Credential provider is optional for client configuration.
            var credentialsProvider = new StaticCredentialsProvider(accessKey, secretKey);
            string endpoints = "foobar.com:8080";

            var producer = new Producer(endpoints)
            {
                CredentialsProvider = credentialsProvider
            };
            string topic = "yourFifoTopic";
            // Set the topic name(s), which is optional. It makes producer could prefetch the topic route before 
            // message publishing.
            producer.AddTopicOfInterest(topic);

            await producer.Start();
            // Define your message body.
            byte[] bytes = Encoding.UTF8.GetBytes("foobar");
            string tag = "yourMessageTagA";
            // You could set multiple keys for the single message.
            var keys = new List<string>
            {
                "yourMessageKey-6cc8b65ed1c8",
                "yourMessageKey-43783375d9a5"
            };
            // Set topic for current message.
            var message = new Message(topic, bytes)
            {
                Tag = tag,
                Keys = keys,
                // Essential for FIFO message, messages that belongs to the same message group follow the FIFO semantics.
                MessageGroup = "yourMessageGroup0"
            };
            var sendReceipt = await producer.Send(message);
            Logger.Info($"Send message successfully, sendReceipt={sendReceipt}.");
            await producer.Shutdown();
        }
    }
}

File: php/examples/ProducerFifoMessageExample.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Producer FIFO Message Example
 * 
 * This example demonstrates how to send FIFO (ordered) messages using RocketMQ PHP SDK.
 * Messages in the same message group will be consumed in order.
 */

require_once __DIR__ . '/../vendor/autoload.php';
require_once __DIR__ . '/../Producer.php';

use Apache\Rocketmq\ClientConfiguration;
use Apache\Rocketmq\Producer;
use Apache\Rocketmq\Builder\MessageBuilder;

// Configuration
$endpoints = '127.0.0.1:8080';
$topic = 'topic-order';
$producerGroup = 'GID-order-consumer';

echo "=== Producer FIFO Message Example ===\n\n";

try {
    // Create client configuration
    $config = new ClientConfiguration($endpoints);
    $config->withSslEnabled(false);
    
    // Create and start producer
    $producer = Producer::getInstance($config, $producerGroup);
    $producer->start();
    echo "✓ Producer started\n\n";
    
    // Send multiple FIFO messages in the same message group
    $messageGroup = 'order_12345';
    
    for ($i = 1; $i <= 5; $i++) {
        $body = "FIFO message #{$i} for order {$messageGroup}";
        $tag = 'fifo-tag';
        $keys = "fifo-key-{$i}";
        
        // Build FIFO message
        $msgBuilder = new MessageBuilder();
        $message = $msgBuilder
            ->setTopic($topic)
            ->setBody($body)
            ->setTag($tag)
            ->setKeys([$keys])
            ->setMessageGroup($messageGroup)  // Important: Set message group for FIFO
            ->build();
        
        // Send message
        $receipt = $producer->send($message);
        echo "✓ Sent FIFO message #{$i}\n";
        echo "  - Message ID: {$receipt->getMessageId()}\n";
        echo "  - Message Group: {$messageGroup}\n";
    }
    
    echo "\nAll messages in the same group will be consumed in order.\n\n";
    
    // Shutdown producer
    $producer->shutdown();
    echo "✓ Producer shutdown\n";
    
} catch (\Exception $e) {
    echo "✗ Error: " . $e->getMessage() . "\n";
    exit(1);
}

echo "\n=== Example Complete ===\n";
